geopandas/io/file.py

```diff
@@ -67,9 +67,14 @@ def read_file(filename, **kwargs):
         else:
             f_filt = f
         gdf = GeoDataFrame.from_features(f_filt, crs=crs)
-        # re-order with column order from metadata, with geometry last
+
         columns = list(f.meta["schema"]["properties"]) + ["geometry"]
-        gdf = gdf[columns]
+        if gdf.empty:
+            # make sure the empty result has the correct column names
+            gdf = GeoDataFrame(columns=columns)
```

Comment:
Isn't the bug really in `from_features()`? Since it is the method that was returning an empty geodataframe?

Reply:
Yes, and I first wanted to fix it there. However, the problem is that `from_features` is more general, and does not only handle fiona collections. A general FeatureCollection (eg from `__geo_interace__`) has no information about the columns if it is empty.